# 项目：评估和清理英国电商公司销售数据

## 分析目标

此数据分析的目的是，根据市场销售数据，挖掘畅销产品，以便制定更有效的市场策略来提升营收。



## 简介

原始数据集记录了一家英国在线零售公司在2010年12月1日至2011年12月9日期间的所有交易情况，涵盖了该公司在全球不同国家和地区的业务数据。该公司主要销售覆盖各个场景的礼品，包括但不限于生日礼品、结婚纪念品、圣诞礼品等等。该公司的客户群体主要包括批发商和个人消费者，其中批发商占据了相当大的比例。

数据每列的含义如下：
- `InvoiceNo`: 发票号码。6位数，作为交易的唯一标识符。如果这个代码以字母“c”开头，表示这笔交易被取消。
- `StockCode`: 产品代码。5位数，作为产品的唯一标识符。
- `Description`: 产品名称。
- `Quantity`: 产品在交易中的数量。
- `InvoiceDate`: 发票日期和时间。交易发生的日期和时间。
- `UnitPrice`: 单价。价格单位为英镑（£）。
- `CustomerID`: 客户编号。5位数，作为客户的唯一标识符。
- `Country`: 国家名称。客户所居住的国家的名称。

## 读取数据

In [8]:
import numpy as np
import pandas as pd
original_data = pd.read_csv("e_commerce.csv")
original_data.sample(10)

InvoiceNo StockCode                        Description  Quantity  \
475068    576860     22988                  SOLDIERS EGG CUP         12   
33600     539267    85099B            JUMBO BAG RED RETROSPOT        10   
141333    548505     82482  WOODEN PICTURE FRAME WHITE FINISH         1   
131220    547549     21559  STRAWBERRY LUNCH BOX WITH CUTLERY         6   
181605    552493     84378      SET OF 3 HEART COOKIE CUTTERS         1   
500797    578808     84879      ASSORTED COLOUR BIRD ORNAMENT        40   
428916    573509     23318     BOX OF 6 MINI VINTAGE CRACKERS        24   
35095     539434     22694                       WICKER STAR          4   
183860    552670     22352  LUNCH BOX WITH CUTLERY RETROSPOT          1   
103392    545062     22429         ENAMEL MEASURING JUG CREAM         1   

             InvoiceDate  UnitPrice  CustomerID         Country  
475068  11/16/2011 15:58       1.25     17602.0  United Kingdom  
33600   12/16/2010 15:23       1.95     13534.0  United Kingdom  
141333   3/31/2011 14:34       4.96         NaN  United Kingdom  
131220   3/23/2011 15:40       2.55     14794.0  United Kingdom  
181605    5/9/2011 16:21       2.46         NaN  United Kingdom  
500797  11/25/2011 13:23       1.69     16327.0  United Kingdom  
428916  10/31/2011 12:21       2.49     16169.0  United Kingdom  
35095   12/17/2010 14:41       4.21         NaN  United Kingdom  
183860   5/10/2011 15:06       4.96         NaN  United Kingdom  
103392   2/27/2011 14:07       4.25     12748.0  United Kingdom

## 评估数据

读取后观察结果：

    1.表格每列表示一个变量，每行表示一个观察值，每个单元格是一个值，不存在显著结构性问题。
    
待评估问题：

    1.是否存在缺失数据（是）。
    
    2.是否存在重复数据。
    
    3.是否存在表述不一致数据。
    
    4.是否存在无效数据。

    5.是否存在异常数据。

    6.是否存在数据类型错误


In [11]:
original_data.info()  # 查看基本信息

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 541909 entries, 0 to 541908
Data columns (total 8 columns):
 #   Column       Non-Null Count   Dtype  
---  ------       --------------   -----  
 0   InvoiceNo    541909 non-null  object 
 1   StockCode    541909 non-null  object 
 2   Description  540455 non-null  object 
 3   Quantity     541909 non-null  int64  
 4   InvoiceDate  541909 non-null  object 
 5   UnitPrice    541909 non-null  float64
 6   CustomerID   406829 non-null  float64
 7   Country      541909 non-null  object 
dtypes: float64(2), int64(1), object(5)
memory usage: 33.1+ MB


#### 评估缺失数据

In [13]:
original_data.isnull().sum() #查看缺失值个数

InvoiceNo           0
StockCode           0
Description      1454
Quantity            0
InvoiceDate         0
UnitPrice           0
CustomerID     135080
Country             0
dtype: int64

从输出结果来看， Description和CustomerID数据存在缺失值，需进一步分析。CustomerID数据类型应为字符串，InvoiceDate的数据类型应为日期，存在数据类型错误。

    1.Description缺失值，根据分析目标，Description表示产品名称，为分析的关键变量，不应该存在缺失值。因此提取缺失数据观察。

    2.CustomerID表示客户编号，不是分析畅销商品的必要变量，因此可以保留此变量为空的观察值。

In [15]:
original_data[original_data["Description"].isnull()]

InvoiceNo StockCode Description  Quantity      InvoiceDate  UnitPrice  \
622       536414     22139         NaN        56  12/1/2010 11:52        0.0   
1970      536545     21134         NaN         1  12/1/2010 14:32        0.0   
1971      536546     22145         NaN         1  12/1/2010 14:33        0.0   
1972      536547     37509         NaN         1  12/1/2010 14:33        0.0   
1987      536549    85226A         NaN         1  12/1/2010 14:34        0.0   
...          ...       ...         ...       ...              ...        ...   
535322    581199     84581         NaN        -2  12/7/2011 18:26        0.0   
535326    581203     23406         NaN        15  12/7/2011 18:31        0.0   
535332    581209     21620         NaN         6  12/7/2011 18:35        0.0   
536981    581234     72817         NaN        27  12/8/2011 10:33        0.0   
538554    581408     85175         NaN        20  12/8/2011 14:06        0.0   

        CustomerID         Country  
622            NaN  United Kingdom  
1970           NaN  United Kingdom  
1971           NaN  United Kingdom  
1972           NaN  United Kingdom  
1987           NaN  United Kingdom  
...            ...             ...  
535322         NaN  United Kingdom  
535326         NaN  United Kingdom  
535332         NaN  United Kingdom  
536981         NaN  United Kingdom  
538554         NaN  United Kingdom  

[1454 rows x 8 columns]

观察结果：缺失Description的交易数据，UnitPrice都为0。为了验证猜想，我们增加筛选条件，看是否存在Description变量缺失且UnitPrice不为0的数据。

In [17]:
original_data[(original_data["Description"].isnull()) & (original_data["UnitPrice"] != 0)]

Empty DataFrame
Columns: [InvoiceNo, StockCode, Description, Quantity, InvoiceDate, UnitPrice, CustomerID, Country]
Index: []

筛选出来结果数量为0条，说明缺失Description值的数据，也同时缺失有效的UnitPrice值。

Description表示产品名称，UnitPrice表示产品单价，都是进行后续商品交易分析的重要变量。如果它们同时缺失，我们认为数据无法提供有效含义，因此这些后续可以被删除。

#### 评估异常数据

In [20]:
original_data.describe()  #查看统计数据

Quantity      UnitPrice     CustomerID
count  541909.000000  541909.000000  406829.000000
mean        9.552250       4.611114   15287.690570
std       218.081158      96.759853    1713.600303
min    -80995.000000  -11062.060000   12346.000000
25%         1.000000       1.250000   13953.000000
50%         3.000000       2.080000   15152.000000
75%        10.000000       4.130000   16791.000000
max     80995.000000   38970.000000   18287.000000

从输出结果来看，Quantity和UnitPrice都存在负数值，需进一步分析确认含义。CustomerID数据类型应为str，存在数据类型错误。

1.筛选Quantity小于0的数据

In [23]:
original_data[original_data["Quantity"] < 0]

InvoiceNo StockCode                       Description  Quantity  \
141      C536379         D                          Discount        -1   
154      C536383    35004C   SET OF 3 COLOURED  FLYING DUCKS        -1   
235      C536391     22556    PLASTERS IN TIN CIRCUS PARADE        -12   
236      C536391     21984  PACK OF 12 PINK PAISLEY TISSUES        -24   
237      C536391     21983  PACK OF 12 BLUE PAISLEY TISSUES        -24   
...          ...       ...                               ...       ...   
540449   C581490     23144   ZINC T-LIGHT HOLDER STARS SMALL       -11   
541541   C581499         M                            Manual        -1   
541715   C581568     21258        VICTORIAN SEWING BOX LARGE        -5   
541716   C581569     84978  HANGING HEART JAR T-LIGHT HOLDER        -1   
541717   C581569     20979     36 PENCILS TUBE RED RETROSPOT        -5   

            InvoiceDate  UnitPrice  CustomerID         Country  
141      12/1/2010 9:41      27.50     14527.0  United Kingdom  
154      12/1/2010 9:49       4.65     15311.0  United Kingdom  
235     12/1/2010 10:24       1.65     17548.0  United Kingdom  
236     12/1/2010 10:24       0.29     17548.0  United Kingdom  
237     12/1/2010 10:24       0.29     17548.0  United Kingdom  
...                 ...        ...         ...             ...  
540449   12/9/2011 9:57       0.83     14397.0  United Kingdom  
541541  12/9/2011 10:28     224.69     15498.0  United Kingdom  
541715  12/9/2011 11:57      10.95     15311.0  United Kingdom  
541716  12/9/2011 11:58       1.25     17315.0  United Kingdom  
541717  12/9/2011 11:58       1.25     17315.0  United Kingdom  

[10624 rows x 8 columns]

从筛选结果来看，Quantity变量为负数的观察值，InvoiceNo似乎均以C开头，表示订单被取消。

为了验证该猜想，增加筛选条件，看是否存在Quantity变量为负数且InvoiceNo不以C开头的观察值。

In [25]:
original_data[(original_data["Quantity"] < 0) & (original_data["InvoiceNo"].str[0] != "C")]

InvoiceNo StockCode Description  Quantity      InvoiceDate  UnitPrice  \
2406      536589     21777         NaN       -10  12/1/2010 16:50        0.0   
4347      536764    84952C         NaN       -38  12/2/2010 14:42        0.0   
7188      536996     22712         NaN       -20  12/3/2010 15:30        0.0   
7189      536997     22028         NaN       -20  12/3/2010 15:30        0.0   
7190      536998     85067         NaN        -6  12/3/2010 15:30        0.0   
...          ...       ...         ...       ...              ...        ...   
535333    581210     23395       check       -26  12/7/2011 18:36        0.0   
535335    581212     22578        lost     -1050  12/7/2011 18:38        0.0   
535336    581213     22576       check       -30  12/7/2011 18:38        0.0   
536908    581226     23090     missing      -338   12/8/2011 9:56        0.0   
538919    581422     23169     smashed      -235  12/8/2011 15:24        0.0   

        CustomerID         Country  
2406           NaN  United Kingdom  
4347           NaN  United Kingdom  
7188           NaN  United Kingdom  
7189           NaN  United Kingdom  
7190           NaN  United Kingdom  
...            ...             ...  
535333         NaN  United Kingdom  
535335         NaN  United Kingdom  
535336         NaN  United Kingdom  
536908         NaN  United Kingdom  
538919         NaN  United Kingdom  

[1336 rows x 8 columns]

以上猜想被证实错误，因为还存在1336条Quantity变量为负数且InvoiceNo不以C开头的观察值。

但根据以上输出结果，这些筛选出的观察值的UnitPrice观察值均为0，因此增加UnitPrice的条件进行验证。

In [27]:
original_data[(original_data["Quantity"] < 0)&(original_data["InvoiceNo"].str[0] != "C")&(original_data["UnitPrice"] != 0)]

Empty DataFrame
Columns: [InvoiceNo, StockCode, Description, Quantity, InvoiceDate, UnitPrice, CustomerID, Country]
Index: []

根据输出结果，猜想得到验证，当Quantity变量为负数的时候，观察值满足以下条件之一：

    InvoiceNo以C开头，表示订单被取消

    UnitPrice为0，说明单价为0英镑

这些交易数据均不是有效成交数据，不贡献销售，不在后续分析范围内，因此我们将在数据清理步骤，将Quantity为负数的观察值删除。

2.筛选UnitPrice小于0的数据

In [30]:
original_data[original_data["UnitPrice"] < 0]

InvoiceNo StockCode      Description  Quantity      InvoiceDate  \
299983   A563186         B  Adjust bad debt         1  8/12/2011 14:51   
299984   A563187         B  Adjust bad debt         1  8/12/2011 14:52   

        UnitPrice  CustomerID         Country  
299983  -11062.06         NaN  United Kingdom  
299984  -11062.06         NaN  United Kingdom

从输出结果来看，UnitPrice为负数的观察值都是坏账调账，不属于实际商品交易数据，因此也在数据清理步骤中也将其删除。

#### 评估重复数据

In [33]:
original_data.duplicated().sum()

5268

从输出结果来看，表格存在5268条重复数据，根据数据变量的含义来看，虽然InvoiceNo、StockCode和CustomerID都是唯一标识符，但一次交易可能包含多件商品，因此InvoiceNo可以存在重复，不同交易可以包含同件商品，因此StockCode可以存在重复，顾客可以进行多次交易或下单多个商品，因此CustomerID也可以存在重复。

那么针对此数据集，我们无需对重复数据进行处理。

#### 评估表述不一致数据

In [36]:
original_data["Country"].value_counts()

Country
United Kingdom          495266
Germany                   9495
France                    8557
EIRE                      8196
Spain                     2533
Netherlands               2371
Belgium                   2069
Switzerland               2002
Portugal                  1519
Australia                 1259
Norway                    1086
Italy                      803
Channel Islands            758
Finland                    695
Cyprus                     622
Sweden                     462
Unspecified                446
Austria                    401
Denmark                    389
Japan                      358
Poland                     341
Israel                     297
China                      288
Singapore                  229
USA                        218
UK                         211
Iceland                    182
Canada                     151
Greece                     146
Malta                      127
United States               73
United Arab Emirates        68


从输出结果来看，"USA"、"United States"均在表示美国，"United Kingdom"、"UK"、"U.K."均在表示英国，因此应该在清洗步骤对这些值进行统一，只保留一个指代值。

#### 评估结论

根据前面评估部分得到的结论，我们需要进行的数据清理包括：

    1.把InvoiceDate变量的数据类型转换为为日期时间
    2.把CustomerID变量的数据类型转换为字符串
    3.把Description变量缺失的观察值删除
    4.把Country变量值"USA"替换为"United States"
    5.把Country变量值"UK"、"U.K."替换为"United Kingdom"
    6.把Quantity变量值为负数的观察值删除
    7.把UnitPrice变量值为负数的观察值删除

## 清理数据

In [41]:
cleaned_data = original_data.copy()
cleaned_data.head()

InvoiceNo StockCode                          Description  Quantity  \
0    536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1    536365     71053                  WHITE METAL LANTERN         6   
2    536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3    536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4    536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   

      InvoiceDate  UnitPrice  CustomerID         Country  
0  12/1/2010 8:26       2.55     17850.0  United Kingdom  
1  12/1/2010 8:26       3.39     17850.0  United Kingdom  
2  12/1/2010 8:26       2.75     17850.0  United Kingdom  
3  12/1/2010 8:26       3.39     17850.0  United Kingdom  
4  12/1/2010 8:26       3.39     17850.0  United Kingdom

1.把InvoiceDate变量的数据类型转换为为日期时间，customerID变量的数据类型转换为字符串(把CustomerID变量值结尾的".0"删除)

In [43]:
cleaned_data["InvoiceDate"] = pd.to_datetime(cleaned_data["InvoiceDate"])
cleaned_data["CustomerID"] = cleaned_data["CustomerID"].astype(str)
cleaned_data["CustomerID"] = cleaned_data["CustomerID"].str.slice(0, -2)
cleaned_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 541909 entries, 0 to 541908
Data columns (total 8 columns):
 #   Column       Non-Null Count   Dtype         
---  ------       --------------   -----         
 0   InvoiceNo    541909 non-null  object        
 1   StockCode    541909 non-null  object        
 2   Description  540455 non-null  object        
 3   Quantity     541909 non-null  int64         
 4   InvoiceDate  541909 non-null  datetime64[ns]
 5   UnitPrice    541909 non-null  float64       
 6   CustomerID   541909 non-null  object        
 7   Country      541909 non-null  object        
dtypes: datetime64[ns](1), float64(1), int64(1), object(5)
memory usage: 33.1+ MB


In [44]:
cleaned_data.head()

InvoiceNo StockCode                          Description  Quantity  \
0    536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1    536365     71053                  WHITE METAL LANTERN         6   
2    536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3    536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4    536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   

          InvoiceDate  UnitPrice CustomerID         Country  
0 2010-12-01 08:26:00       2.55      17850  United Kingdom  
1 2010-12-01 08:26:00       3.39      17850  United Kingdom  
2 2010-12-01 08:26:00       2.75      17850  United Kingdom  
3 2010-12-01 08:26:00       3.39      17850  United Kingdom  
4 2010-12-01 08:26:00       3.39      17850  United Kingdom

2.把Description变量缺失的观察值删除

In [46]:
cleaned_data.dropna(subset=["Description"], inplace=True)
cleaned_data["Description"].isnull().sum()

0

3.把Country变量值"USA"替换为"United States", 把Country变量值"UK"、"U.K."替换为"United Kingdom"

In [48]:
cleaned_data["Country"] = cleaned_data["Country"].replace({"USA": "United States"})
cleaned_data["Country"] = cleaned_data["Country"].replace({"UK": "United Kingdom", "U.K.": "United Kingdom"})

In [49]:
print(len(cleaned_data[cleaned_data["Country"] == "USA"]))
print(len(cleaned_data[cleaned_data["Country"] == "UK"]))
print(len(cleaned_data[cleaned_data["Country"] == "U.K."]))

0
0
0


4.把Quantity变量值为负数的观察值和UnitPrice变量值为负数的观察值删除

In [51]:
cleaned_data = cleaned_data[cleaned_data["Quantity"] >= 0]
cleaned_data = cleaned_data[cleaned_data["UnitPrice"] >= 0]
print(len(cleaned_data[cleaned_data["Quantity"] < 0]))
print(len(cleaned_data[cleaned_data["UnitPrice"] < 0]))

0
0


## 保存清理后的数据

In [53]:
cleaned_data["Sales"]=cleaned_data["Quantity"] * cleaned_data["UnitPrice"]
cleaned_data.head(10)

InvoiceNo StockCode                          Description  Quantity  \
0    536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1    536365     71053                  WHITE METAL LANTERN         6   
2    536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3    536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4    536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   
5    536365     22752         SET 7 BABUSHKA NESTING BOXES         2   
6    536365     21730    GLASS STAR FROSTED T-LIGHT HOLDER         6   
7    536366     22633               HAND WARMER UNION JACK         6   
8    536366     22632            HAND WARMER RED POLKA DOT         6   
9    536367     84879        ASSORTED COLOUR BIRD ORNAMENT        32   

          InvoiceDate  UnitPrice CustomerID         Country  Sales  
0 2010-12-01 08:26:00       2.55      17850  United Kingdom  15.30  
1 2010-12-01 08:26:00       3.39      17850  United Kingdom  20.34  
2 2010-12-01 08:26:00       2.75      17850  United Kingdom  22.00  
3 2010-12-01 08:26:00       3.39      17850  United Kingdom  20.34  
4 2010-12-01 08:26:00       3.39      17850  United Kingdom  20.34  
5 2010-12-01 08:26:00       7.65      17850  United Kingdom  15.30  
6 2010-12-01 08:26:00       4.25      17850  United Kingdom  25.50  
7 2010-12-01 08:28:00       1.85      17850  United Kingdom  11.10  
8 2010-12-01 08:28:00       1.85      17850  United Kingdom  11.10  
9 2010-12-01 08:34:00       1.69      13047  United Kingdom  54.08

In [54]:
cleaned_data.to_csv("e_commerce_cleaned.csv", index=False)